Plutarch/DataRepr/Internal.hs

```diff
@@ -180,7 +184,55 @@ type family PUnLabel (a :: PLabeledType) :: PType where
  This is how most data structures are stored on-chain.
 -}
 type PDataSum :: [[PLabeledType]] -> PType
-data PDataSum (defs :: [[PLabeledType]]) (s :: S)
+data PDataSum (defs :: [[PLabeledType]]) (s :: S) where 
+  DSZ :: (forall {out :: PType}. (Term s (PDataRecord x) -> Term s out) -> Term s out) 
+      -> PDataSum (x ': xs) s
+
+  DSS :: PDataSum xs s -> PDataSum (x ': xs) s
+
+-- the length functions in sop-core gives the actual length, we want length-1
+adjustedLength :: forall x xs proxy. SListI (x:xs) => proxy (x:xs) -> Integer 
+adjustedLength proxy = fromIntegral (lengthSList proxy) - 1  
+
+instance SListI (def:defs) => PlutusType (PDataSum (def:defs)) where
+  type PInner (PDataSum (def:defs)) _ = PData 
+
+  pcon' :: forall 
+           {s :: S} 
+           {b :: PType}
+         . PDataSum (def:defs) s 
+        -> Term s (PInner (PDataSum (def:defs)) b)
+  pcon'  = \case 
+        DSZ f ->  punsafeBuiltin PLC.ConstrData # (pconstant @PInteger $ adjustedLength $ Proxy @(def:defs)) # (f id) -- rewrite this with pconstrBuiltin?
+        DSS rest -> case rest of 
+          DSZ f -> punsafeBuiltin PLC.ConstrData # (pconstant @PInteger $ adjustedLength $ Proxy @(def:defs)) # (f id)
+          dss@(DSS _) -> pcon' dss  
+
+  pmatch' :: forall 
+             {s :: S} 
+             {b :: PType}
+           . Term s (PInner (PDataSum (def:defs)) b)
+          -> (PDataSum (def:defs) s -> Term s b) 
+          -> Term s b
+  pmatch' d f = 
```

Comment:
Can we use `pmatchDataSum` here somehow?

Reply:
I don't  _think_ we can. But I could be wrong and if it's important I can look into it more. 

If I remember right the problem is that we can't get the `Nat` we need to use `mkDataReprHandler` from the `PInteger` that indicates the index of the constructor, and we can't construct the handlers manually because there's no (straightforward) way (that I could think of) to get a `Term s (PDataRecord def) -> Term s out`
from a `PDataSum defs s -> Term s out`. I'm less confident that there's no way around the second problem  -- there could well be something in the rest of the library that I'm not aware of which would make that possible.  

Does that make sense? If my reasoning here is flawed let me know.

Reply:
I very much feel like we can use `pmatchDataSum` here.

We can have a type which internal representation is `PDataSum defs`, and implements `PlutusType` such that `pmatch` uses `pmatchDataSum`, so we should be able to do the same here. `pmatchDataSum` implements important optimisations. Look at the `PIsDataRepr` class and generic instance for `PlutusType` using the `deriving via` wrapper.

Reply:
Ok, I will take another look at it when I have some time. I did not look very closely at `PIsDataRepr`, hopefully that opens up some possibility I wasn't seeing. 

Just a heads up that I might not be able to get to this for a day or two.